File: packaging/install/chorus_executor.rb
require_relative 'installer_errors'

class ChorusExecutor
  attr_writer :destination_path, :version

  def initialize(options = {})
    @logger = options[:logger]
    @debug = options[:debug] == true

    raise InstallerErrors::InstallationFailed.new("Logger must be set") unless @logger
  end

  def exec(command)
    full_command = "PATH=#{release_path}/postgres/bin:$PATH && #{command}"
    @logger.debug(full_command)
    @logger.capture_output(full_command) || raise(InstallerErrors::CommandFailed, command)
  end

  def rake(command)
    exec "cd #{release_path} && RAILS_ENV=production bin/rake #{command}#{if_debug(' --trace')}"
  end

  def start_chorus
    chorus_control "start"
  end

  def start_postgres
    @logger.log "starting postgres..."
    chorus_control "start postgres"
  end

  def stop_postgres
    if File.directory? "#{release_path}/postgres"
      @logger.log "stopping postgres..."
      chorus_control "stop postgres"
    end
  end

  def initdb(data_path, database_user)
    exec "initdb --locale=en_US.UTF-8 -D #{data_path}/db --auth=md5 --pwfile=#{release_path}/postgres/pwfile --username=#{database_user}"
  end

  def extract_postgres(package_name)
    exec "tar xzf#{if_debug("v")} #{release_path}/packaging/postgres/#{package_name} -C #{release_path}"
  end

  def start_previous_release
    previous_chorus_control "start"
  end

  def stop_previous_release
    previous_chorus_control "stop"
  end

  def import_legacy_schema(legacy_installation_path)
    exec "cd #{release_path} && INSTALL_ROOT=#{@destination_path} CHORUS_HOME=#{release_path} packaging/chorus_migrate -s legacy_database.sql -w #{legacy_installation_path}/chorus-apps/runtime/data"
  end

  def dump_legacy_data
    exec "cd #{release_path} && PGUSER=edcadmin pg_dump -p 8543 chorus -O -f legacy_database.sql"
  end

  def stop_legacy_app(legacy_installation_path)
    legacy_exec legacy_installation_path, "bin/edcsvrctl stop; true"
  end

  def stop_legacy_app!(legacy_installation_path)
    legacy_exec legacy_installation_path, "bin/edcsvrctl stop"
  end

  def start_legacy_postgres(legacy_installation_path)
    legacy_exec legacy_installation_path, "bin/edcsvrctl start || bin/edcsvrctl start"
  end

  private

  def release_path
    "#{@destination_path}/releases/#{@version}"
  end

  def legacy_exec(legacy_installation_path, command)
    Dir.chdir(legacy_installation_path) do
      exec "source #{legacy_installation_path}/edc_path.sh && #{command}"
    end
  end

  def chorus_control(command)
    exec "CHORUS_HOME=#{release_path} #{release_path}/packaging/chorus_control.sh #{command}"
  end

  def previous_chorus_control(command)
    exec "CHORUS_HOME=#{@destination_path}/current #{@destination_path}/chorus_control.sh #{command}"
  end

  def if_debug(arg)
    if @debug
      arg
    else
      ""
    end
  end
end

File: packaging/setup/chorus_executor.py
import os
import pwd
import subprocess
import shlex
from log import logger
from helper import get_version
class PSQLException(Exception):
    pass

class RAKEException(Exception):
    pass

class ChorusExecutor:
    def __init__(self, chorus_path=None):
        self.chorus_path = chorus_path
        if chorus_path is None:
            self.release_path = None
        else:
            self.release_path = os.path.join(chorus_path, 'releases/%s' % get_version(chorus_path))

    def call(self, command):
        logger.debug(command)
        return subprocess.call(shlex.split(command))

    def run(self, command, postgres_bin_path=None):
        if self.release_path is not None:
            if postgres_bin_path is None:
                postgres_bin_path = self.release_path
            command = "PATH=%s/postgres/bin:$PATH && %s" % (postgres_bin_path, command)

        logger.debug(command)
        p = subprocess.Popen(command, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
        stdout, stderr = p.communicate()
        if stdout:
            logger.debug(stdout)
        if stderr:
            logger.debug(stderr)
        return p.returncode, stdout, stderr

    def run_as_user(self, command, user):
        def demote(user_uid, user_gid):
            os.setgid(user_gid)
            os.setuid(user_uid)
        pw_record = pwd.getpwnam(user)
        user_uid = pw_record.pw_uid
        user_gid = pw_record.pw_gid
        user_name = pw_record.pw_name
        user_home_dir = pw_record.pw_dir
        env = os.environ.copy()
        env[ 'HOME'     ] = user_home_dir
        env[ 'LOGNAME'  ] = user_name
        env[ 'USER'     ] = user_name
        return subprocess.call(shlex.split(command), preexec_fn=demote(user_uid, user_gid), env=env)

    def extract_postgres(self, package_name):
        self.run("tar xzfv %s -C %s" % (os.path.join(self.release_path, "packaging/postgres/" + package_name),\
                                        self.release_path))

    def chorus_control(self, command):
       command = "CHORUS_HOME=%s %s %s/packaging/chorus_control.sh %s" % \
               (self.release_path, self.alpine_env(), self.release_path, command)
       (ret, stdout, stderr) = self.run(command)
       if ret != 0:
           raise Exception("chorus_constrol.sh command failed")
       return (ret, stdout, stderr)

    def previous_chorus_control(self, command):
        command = "CHORUS_HOME=%s %s %s %s" % \
                (os.path.join(self.chorus_path, "current"), self.alpine_env(),\
                os.path.join(self.chorus_path, "chorus_control.sh"), command)
        (ret, stdout, stderr) = self.run(command)
        if ret != 0:
           raise Exception("chorus_constrol.sh command failed")
        return (ret, stdout, stderr)

    def alpine_env(self):
        return "ALPINE_HOME=%s/alpine-current ALPINE_DATA_REPOSITORY=%s/shared/ALPINE_DATA_REPOSITORY" % \
                (self.chorus_path, self.chorus_path)

    def start_previous_release(self):
        self.previous_chorus_control("start")

    def stop_previous_release(self):
        self.previous_chorus_control("stop")
        #self.run("killall chorus")
    def start_postgres(self):
        logger.debug("Starting postgres...")
        ret, stdout, stderr = self.chorus_control("start postgres")
        if "postgres failed" in stdout:
            raise PSQLException(stdout)

    def stop_postgres(self):
        logger.debug("Stopping postgres")
        ret, stdout, stderr = self.chorus_control("stop postgres")
        if "postgres failed" in stdout:
            raise PSQLException(stdout)

    def initdb(self, data_path, database_user):
        command = "initdb --locale=en_US.UTF-8 -D %s/db --auth=md5 --pwfile=%s/postgres/pwfile --username=%s" % \
                (data_path, self.release_path, database_user)
        ret, stdout, stderr = self.run(command)
        #if "exists but is not empty" in stderr:
        #    logger.warning(stderr)

    def rake(self, command):
        command = "cd %s && RAILS_ENV=production bin/ruby -S bin/rake %s --trace" % \
                (self.release_path, command)
        ret, stdout, stderr = self.run(command)
        if "rake aborted" in stderr:
            raise RAKEException(stderr)

